### ISSUE TITLE
win_environment changes don't take into effect until after user logout

### ISSUE THREAD

**Issue by synapticarbors** [ansible/ansible-modules-extras#908]
Issue Type:
- Bug Report

Ansible Version: 

```
ansible 1.9.2
  configured module search path = None
```

I attempted to test this off of the latest `devel` branch, but I was getting WinRM ssl errors that weren't resolved using the callback plugin that has been mentioned on the Google Group, and which works for `v1.9.2`.

Ansible Configuration:

```
[defaults]
callback_plugins=<path stuff>/test-ansible/callback_plugins
```

Environment:

Running Ansible off of OSX 10.9.6, targeting a Windows server 2012 R2 

Summary:

The `win_environment` module appears to successfully set an environment variable, but the env variable is not visible on the target machine until the user logs out and then logs back in. 

Steps To Reproduce:

I copied the `win_environment` module (both `.py` and `.ps1` files) into `./library` since they are only available in v2 (see above). I then use that module in a task to set an environment variable as:

```

---
- name: Test ansible
  hosts: windows
  remote_user: deploy
  tasks:
    - name: Set ABC environment variable
      win_environment:
          state: present
          name: 'ABC'
          value: "{{ abc }}"
          level: User
```

Expected Results:

I expect that given a fresh powershell that `$env:ABC` will be the value provided in the playbook. 

Actual Results:

If `ABC` is not set on the target machine, I get a result as:

```
ok: [win_machine] => {"before_value": "null", "changed": true, "level": "user", "name": "ABC", "value": "xxx"}
```

Switching over to an RDP on `win_machine`, `ABC` is not set. Re-running the ansible play, I get:

```
ok: [win_machine] => {"before_value": "xxx", "changed": false, "level": "user", "name": "ABC", "value": "xxx"}
```

so, the new value is being stored somewhere. 

If on the Windows machine I run either `setx ABC yyy` from a command prompt or `[Environment]::SetEnvironmentVariable('ABC', 'yyy', 'User')` from the powershell, and then re-run the ansible task on my OSX machine, I get:

```
ok: [win_machine] => {"before_value": "yyy", "changed": true, "level": "user", "name": "ABC", "value": "xxx"}
```

However, again, the change isn't reflected on the machine after starting up a new powershell terminal. The only way to get the value pushed by Ansible to be reflected on the user machine is to log the user off and then log them back on. This doesn't seem to be the correct behavior, or at least not what I'd anticipate given how running similar commands locally on the target machine in a script behaves. 

It seems like the Windows machine is holding the values that Ansible pushed in some sort of temporary registry that doesn't get flushed until a logout. Restarting the WinRM service doesn't seem to effect things. I can run a script on the Windows machine that just contains:

```
[Environment]::SetEnvironmentVariable('ABC', 'xxx', 'User')
```

and that takes hold immediately once you fire up a new powershell.

**Comment by jhawkesworth** [comment]
Interesting... my playbooks haven't had cause to use User level yet.
Did a bit of digging and I think it may need a code change to broadcast a setting change event - see https://social.msdn.microsoft.com/Forums/windows/en-US/ce540c7d-a113-4f39-956e-0af6bc91abd3/force-updaterefresh-of-windows-start-menu?forum=winforms 

However, it looks like its up to other applications to notice and react to the setting change event so won't know till its implemented if it will fix.

**Comment by jhawkesworth** [comment]
Working on adding the setting change notification.  Once I have the necessary c# compiling I will share a new version of win_envionment.ps1

**Comment by jhawkesworth** [comment]
Ok I implemented the setting change notification, but on the quick test I did, it doesn't make any difference, the behaviour is still as you described.

I didn't try restarting winrm or wmf to see if that made a difference though.  If you want to give it a try I'll post the modified win_environment.ps1 below.

Otherwise I think best option is to document more clearly.

**Comment by jhawkesworth** [comment]
``` powershell
#!powershell
# This file is part of Ansible
#
# Copyright 2015, Jon Hawkesworth (@jhawkesworth) <figs@unity.demon.co.uk>
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

# WANT_JSON
# POWERSHELL_COMMON

# credits: http://blogs.technet.com/b/stefan_gossner/archive/2010/05/07/using-csharp-c-code-in-powershell-scripts.aspx
# https://social.msdn.microsoft.com/Forums/windows/en-US/ce540c7d-a113-4f39-956e-0af6bc91abd3/force-updaterefresh-of-windows-start-menu?forum=winforms

$scn_class=@"

using System.Text;
using System;
using System.Runtime.InteropServices;

public static class SettingChangeNotifier
 {
  [DllImport("user32.dll", SetLastError = true)]
  private static extern IntPtr SendMessageTimeout ( IntPtr hWnd, int Msg, IntPtr wParam, string lParam, uint fuFlags, uint uTimeout, IntPtr lpdwResult );
  private static readonly IntPtr HWND_BROADCAST = new IntPtr(0xffff);
  private const int WM_SETTINGCHANGE = 0x1a;
  private const int SMTO_ABORTIFHUNG = 0x0002;

  public static void notify ()
  {
   SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE, IntPtr.Zero, null, SMTO_ABORTIFHUNG, 100, IntPtr.Zero);
  }
 }
"@

Add-Type -TypeDefinition $scn_class

$params = Parse-Args $args;
$result = New-Object PSObject;
Set-Attr $result "changed" $false;

If ($params.state) {
    $state = $params.state.ToString().ToLower()
    If (($state -ne 'present') -and ($state -ne 'absent') ) {
        Fail-Json $result "state is '$state'; must be 'present', or 'absent'"
    }
} else {
    $state = 'present'
}

If ($params.name)
{
    $name = $params.name
} else {
    Fail-Json $result "missing required argument: name"
}

$value = $params.value

If ($params.level) {
    $level = $params.level.ToString().ToLower()
    If (( $level -ne 'machine') -and ( $level -ne 'user' ) -and ( $level -ne 'process')) {
        Fail-Json $result "level is '$level'; must be 'machine', 'user', or 'process'"
    }
}

$before_value = [Environment]::GetEnvironmentVariable($name, $level)

if ($state -eq "present" ) {
   [Environment]::SetEnvironmentVariable($name, $value, $level)
   [SettingChangeNotifier]::notify()
} Elseif ($state -eq "absent") {
   [Environment]::SetEnvironmentVariable($name, $null, $level)
   [SettingChangeNotifier]::notify()
}

$after_value = [Environment]::GetEnvironmentVariable($name, $level)

Set-Attr $result "name" $name;
Set-Attr $result "before_value" $before_value;
Set-Attr $result "value" $after_value;
Set-Attr $result "level" $level;
if ($before_value -ne $after_value) {
   Set-Attr $result "changed" $true;
}

Exit-Json $result;
```

**Comment by jhawkesworth** [comment]
I have not been able to find a way to change the described behaviour.  I will create a documentation PR to clarify what the module can do.

**Comment by jhawkesworth** [comment]
Hi, @synapticarbors if you think the changes I have made to the documentation in https://github.com/ansible/ansible-modules-extras/pull/1066 cover the current behaviour, please can this issue be closed?

**Comment by synapticarbors** [comment]
@jhawkesworth that language seems reasonable. Unfortunately I ended up bailing on Ansible since the ease of doing things with Windows was lacking in a way that met my constraints. I ended up going with a local solution for now. 

I'm not sure if it's helpful in the context of Ansible, but the code that Chocolatey uses to update its env variables in session is here:
https://github.com/chocolatey/chocolatey/blob/master/src/helpers/functions/Update-SessionEnvironment.ps1

Thanks for looking into this.
